language/vm/src/normalized.rs

```diff
@@ -0,0 +1,288 @@
+    },
+};
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::Identifier,
+    language_storage::{StructTag, TypeTag},
+};
+
+/// Defines normalized representations of Move types, fields, kinds, structs, functions, and
+/// modules. These representations are useful in situations that require require comparing
+/// functions, resources, and types across modules. This arises in linking, compatibility checks
+/// (e.g., "is it safe to deploy this new module without updating its dependents and/or restarting
+/// genesis?"), defining schemas for resources stored on-chain, and (possibly in the future)
+/// allowing module updates transactions.
+
+/// A normalized version of `SignatureToken`, a type expression appearing in struct or function
+/// declarations. Unlike `SignatureToken`s, `normalized::Type`s from different modules can safely be
+/// compared.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Type {
+    Bool,
+    U8,
+    U64,
+    U128,
+    Address,
+    Signer,
+    Struct {
+        address: AccountAddress,
+        module: Identifier,
+        name: Identifier,
+        type_arguments: Vec<Type>,
+    },
+    Vector(Box<Type>),
+    TypeParameter(TypeParameterIndex),
+    Reference(Box<Type>),
+    MutableReference(Box<Type>),
+}
+
+/// Normalized version of a `FieldDefinition`. The `name` is included even though it is
+/// metadata that it is ignored by the VM. The reason: names are important to clients. We would
+/// want a change from `Account { bal: u64, seq: u64 }` to `Account { seq: u64, bal: u64 }` to be
+/// marked as incompatible. Not safe to compare without an enclosing `Struct`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Field {
+    pub name: Identifier,
+    pub type_: Type,
+}
+
+/// Normalized version of a kind constraint on a generic type parameter. At present, this is
+/// identical to file_format::Kind, though this may change in the future.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Kind {
+    All,
+    Resource,
+    Copyable,
+}
+
+/// Normalized version of a `StructDefinition`. Not safe to compare without an associated
+/// `ModuleId` or `Module`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Struct {
+    pub name: Identifier,
+    pub kind: Kind,
+    pub type_parameters: Vec<Kind>,
+    pub fields: Vec<Field>,
+}
+
+/// Normalized version of a `FunctionDefinition`. Not safe to compare without an associated
+/// `ModuleId` or `Module`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FunctionSignature {
+    pub name: Identifier,
+    pub type_parameters: Vec<Kind>,
+    pub formals: Vec<Type>,
+    pub ret: Vec<Type>,
+}
+
+/// Normalized version of a `CompiledModule`: its address, name, struct declarations, and public
+/// function declarations.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Module {
+    pub address: AccountAddress,
+    pub name: Identifier,
+    pub structs: Vec<Struct>,
+    pub public_functions: Vec<FunctionSignature>,
```

Comment:
I know right now we only need the function signatures, but why not just include all function information here? 

If we do this, we could maybe in the future rewrite the bytecode verifier to work over this "normalized" version rather than having to deal with the soup of indexes in the file format

Reply:
What other normalized function information does the bytecode verifier need? I would expect it to only need normalized function signatures and struct handles in the linking phase and for every other check to be intra-module (and thus not require a normalized representation).

Reply:
I was imagining something a lot bigger than what is currently here, which is really a complete AST-style transformation over the file format, so we'd basically inline all of the indexes with their target (whether it be via copying or by reference or some other memory indirection magic)

This was going back to what @dariorussi was saying in the meeting of having a one-stop-shop for writing tools over the file_format. But perhaps that is outside the scope of this PR

Reply:
I think this is very useful to have, but definitely not the goal of this PR. @emmazzz's stackless bytecode (or a variant) of it seems like what you want here--it normalizes away indexes and the operand stack and some other nasty stuff too.